File: lua/codetyper/features/ask/intent.lua
---@mod codetyper.ask.intent Intent detection for Ask mode
---@brief [[
--- Analyzes user prompts to detect intent (ask/explain vs code generation).
--- Routes to appropriate prompt type and context sources.
---@brief ]]

local M = {}

---@alias IntentType "ask"|"explain"|"generate"|"refactor"|"document"|"test"

---@class Intent
---@field type IntentType Detected intent type
---@field confidence number 0-1 confidence score
---@field needs_project_context boolean Whether project-wide context is needed
---@field needs_brain_context boolean Whether brain/learned context is helpful
---@field needs_exploration boolean Whether full project exploration is needed
---@field keywords string[] Keywords that influenced detection

--- Patterns for detecting ask/explain intent (questions about code)
local ASK_PATTERNS = {
	-- Question words
	{ pattern = "^what%s", weight = 0.9 },
	{ pattern = "^why%s", weight = 0.95 },
	{ pattern = "^how%s+does", weight = 0.9 },
	{ pattern = "^how%s+do%s+i", weight = 0.7 }, -- Could be asking for code
	{ pattern = "^where%s", weight = 0.85 },
	{ pattern = "^when%s", weight = 0.85 },
	{ pattern = "^which%s", weight = 0.8 },
	{ pattern = "^who%s", weight = 0.85 },
	{ pattern = "^can%s+you%s+explain", weight = 0.95 },
	{ pattern = "^could%s+you%s+explain", weight = 0.95 },
	{ pattern = "^please%s+explain", weight = 0.95 },

	-- Explanation requests
	{ pattern = "explain%s", weight = 0.9 },
	{ pattern = "describe%s", weight = 0.85 },
	{ pattern = "tell%s+me%s+about", weight = 0.85 },
	{ pattern = "walk%s+me%s+through", weight = 0.9 },
	{ pattern = "help%s+me%s+understand", weight = 0.95 },
	{ pattern = "what%s+is%s+the%s+purpose", weight = 0.95 },
	{ pattern = "what%s+does%s+this", weight = 0.9 },
	{ pattern = "what%s+does%s+it", weight = 0.9 },
	{ pattern = "how%s+does%s+this%s+work", weight = 0.95 },
	{ pattern = "how%s+does%s+it%s+work", weight = 0.95 },

	-- Understanding queries
	{ pattern = "understand", weight = 0.7 },
	{ pattern = "meaning%s+of", weight = 0.85 },
	{ pattern = "difference%s+between", weight = 0.9 },
	{ pattern = "compared%s+to", weight = 0.8 },
	{ pattern = "vs%s", weight = 0.7 },
	{ pattern = "versus", weight = 0.7 },
	{ pattern = "pros%s+and%s+cons", weight = 0.9 },
	{ pattern = "advantages", weight = 0.8 },
	{ pattern = "disadvantages", weight = 0.8 },
	{ pattern = "trade%-?offs?", weight = 0.85 },

	-- Analysis requests
	{ pattern = "analyze", weight = 0.85 },
	{ pattern = "review", weight = 0.7 }, -- Could also be refactor
	{ pattern = "overview", weight = 0.9 },
	{ pattern = "summary", weight = 0.9 },
	{ pattern = "summarize", weight = 0.9 },

	-- Question marks (weaker signal)
	{ pattern = "%?$", weight = 0.3 },
	{ pattern = "%?%s*$", weight = 0.3 },
}

--- Patterns for detecting code generation intent
local GENERATE_PATTERNS = {
	-- Direct commands
	{ pattern = "^create%s", weight = 0.9 },
	{ pattern = "^make%s", weight = 0.85 },
	{ pattern = "^build%s", weight = 0.85 },
	{ pattern = "^write%s", weight = 0.9 },
	{ pattern = "^add%s", weight = 0.85 },
	{ pattern = "^implement%s", weight = 0.95 },
	{ pattern = "^generate%s", weight = 0.95 },
	{ pattern = "^code%s", weight = 0.8 },

	-- Modification commands
	{ pattern = "^fix%s", weight = 0.9 },
	{ pattern = "^change%s", weight = 0.8 },
	{ pattern = "^update%s", weight = 0.75 },
	{ pattern = "^modify%s", weight = 0.8 },
	{ pattern = "^replace%s", weight = 0.85 },
	{ pattern = "^remove%s", weight = 0.85 },
	{ pattern = "^delete%s", weight = 0.85 },

	-- Feature requests
	{ pattern = "i%s+need%s+a", weight = 0.8 },
	{ pattern = "i%s+want%s+a", weight = 0.8 },
	{ pattern = "give%s+me", weight = 0.7 },
	{ pattern = "show%s+me%s+how%s+to%s+code", weight = 0.9 },
	{ pattern = "how%s+do%s+i%s+implement", weight = 0.85 },
	{ pattern = "can%s+you%s+write", weight = 0.9 },
	{ pattern = "can%s+you%s+create", weight = 0.9 },
	{ pattern = "can%s+you%s+add", weight = 0.85 },
	{ pattern = "can%s+you%s+make", weight = 0.85 },

	-- Code-specific terms
	{ pattern = "function%s+that", weight = 0.85 },
	{ pattern = "class%s+that", weight = 0.85 },
	{ pattern = "method%s+that", weight = 0.85 },
	{ pattern = "component%s+that", weight = 0.85 },
	{ pattern = "module%s+that", weight = 0.85 },
	{ pattern = "api%s+for", weight = 0.8 },
	{ pattern = "endpoint%s+for", weight = 0.8 },
}

--- Patterns for detecting refactor intent
local REFACTOR_PATTERNS = {
	{ pattern = "^refactor%s", weight = 0.95 },
	{ pattern = "refactor%s+this", weight = 0.95 },
	{ pattern = "clean%s+up", weight = 0.85 },
	{ pattern = "improve%s+this%s+code", weight = 0.85 },
	{ pattern = "make%s+this%s+cleaner", weight = 0.85 },
	{ pattern = "simplify", weight = 0.8 },
	{ pattern = "optimize", weight = 0.75 }, -- Could be explain
	{ pattern = "reorganize", weight = 0.9 },
	{ pattern = "restructure", weight = 0.9 },
	{ pattern = "extract%s+to", weight = 0.9 },
	{ pattern = "split%s+into", weight = 0.85 },
	{ pattern = "dry%s+this", weight = 0.9 }, -- Don't repeat yourself
	{ pattern = "reduce%s+duplication", weight = 0.9 },
}

--- Patterns for detecting documentation intent
local DOCUMENT_PATTERNS = {
	{ pattern = "^document%s", weight = 0.95 },
	{ pattern = "add%s+documentation", weight = 0.95 },
	{ pattern = "add%s+docs", weight = 0.95 },
	{ pattern = "add%s+comments", weight = 0.9 },
	{ pattern = "add%s+docstring", weight = 0.95 },
	{ pattern = "add%s+jsdoc", weight = 0.95 },
	{ pattern = "write%s+documentation", weight = 0.95 },
	{ pattern = "document%s+this", weight = 0.95 },
}

--- Patterns for detecting test generation intent
local TEST_PATTERNS = {
	{ pattern = "^test%s", weight = 0.9 },
	{ pattern = "write%s+tests?%s+for", weight = 0.95 },
	{ pattern = "add%s+tests?%s+for", weight = 0.95 },
	{ pattern = "create%s+tests?%s+for", weight = 0.95 },
	{ pattern = "generate%s+tests?", weight = 0.95 },
	{ pattern = "unit%s+tests?", weight = 0.9 },
	{ pattern = "test%s+cases?%s+for", weight = 0.95 },
	{ pattern = "spec%s+for", weight = 0.85 },
}

--- Patterns indicating project-wide context is needed
local PROJECT_CONTEXT_PATTERNS = {
	{ pattern = "project", weight = 0.9 },
	{ pattern = "codebase", weight = 0.95 },
	{ pattern = "entire", weight = 0.7 },
	{ pattern = "whole", weight = 0.7 },
	{ pattern = "all%s+files", weight = 0.9 },
	{ pattern = "architecture", weight = 0.95 },
	{ pattern = "structure", weight = 0.85 },
	{ pattern = "how%s+is%s+.*%s+organized", weight = 0.95 },
	{ pattern = "where%s+is%s+.*%s+defined", weight = 0.9 },
	{ pattern = "dependencies", weight = 0.85 },
	{ pattern = "imports?%s+from", weight = 0.7 },
	{ pattern = "modules?", weight = 0.6 },
	{ pattern = "packages?", weight = 0.6 },
}

--- Patterns indicating project exploration is needed (full indexing)
local EXPLORE_PATTERNS = {
	{ pattern = "explain%s+.*%s*project", weight = 1.0 },
	{ pattern = "explain%s+.*%s*codebase", weight = 1.0 },
	{ pattern = "explain%s+me%s+the%s+project", weight = 1.0 },
	{ pattern = "tell%s+me%s+about%s+.*%s*project", weight = 0.95 },
	{ pattern = "what%s+is%s+this%s+project", weight = 0.95 },
	{ pattern = "overview%s+of%s+.*%s*project", weight = 0.95 },
	{ pattern = "understand%s+.*%s*project", weight = 0.9 },
	{ pattern = "analyze%s+.*%s*project", weight = 0.9 },
	{ pattern = "explore%s+.*%s*project", weight = 1.0 },
	{ pattern = "explore%s+.*%s*codebase", weight = 1.0 },
	{ pattern = "index%s+.*%s*project", weight = 1.0 },
	{ pattern = "scan%s+.*%s*project", weight = 0.95 },
}

--- Match patterns against text
---@param text string Lowercased text to match
---@param patterns table Pattern list with weights
---@return number Score, string[] Matched keywords
local function match_patterns(text, patterns)
	local score = 0
	local matched = {}

	for _, p in ipairs(patterns) do
		if text:match(p.pattern) then
			score = score + p.weight
			table.insert(matched, p.pattern)
		end
	end

	return score, matched
end

--- Detect intent from user prompt
---@param prompt string User's question/request
---@return Intent Detected intent
function M.detect(prompt)
	local text = prompt:lower()

	-- Calculate raw scores for each intent type (sum of matched weights)
	local ask_score, ask_kw = match_patterns(text, ASK_PATTERNS)
	local gen_score, gen_kw = match_patterns(text, GENERATE_PATTERNS)
	local ref_score, ref_kw = match_patterns(text, REFACTOR_PATTERNS)
	local doc_score, doc_kw = match_patterns(text, DOCUMENT_PATTERNS)
	local test_score, test_kw = match_patterns(text, TEST_PATTERNS)
	local proj_score, _ = match_patterns(text, PROJECT_CONTEXT_PATTERNS)
	local explore_score, _ = match_patterns(text, EXPLORE_PATTERNS)

	-- Find the winner by raw score (highest accumulated weight)
	local scores = {
		{ type = "ask", score = ask_score, keywords = ask_kw },
		{ type = "generate", score = gen_score, keywords = gen_kw },
		{ type = "refactor", score = ref_score, keywords = ref_kw },
		{ type = "document", score = doc_score, keywords = doc_kw },
		{ type = "test", score = test_score, keywords = test_kw },
	}

	table.sort(scores, function(a, b)
		return a.score > b.score
	end)

	local winner = scores[1]

	-- If top score is very low, default to ask (safer for Q&A)
	if winner.score < 0.3 then
		winner = { type = "ask", score = 0.5, keywords = {} }
	end

	-- If ask and generate are close AND there's a question mark, prefer ask
	if winner.type == "generate" and ask_score > 0 then
		if text:match("%?%s*$") and ask_score >= gen_score * 0.5 then
			winner = { type = "ask", score = ask_score, keywords = ask_kw }
		end
	end

	-- Determine if "explain" vs "ask" (explain needs more context)
	local intent_type = winner.type
	if intent_type == "ask" then
		-- "explain" if asking about how something works, otherwise "ask"
		if text:match("explain") or text:match("how%s+does") or text:match("walk%s+me%s+through") then
			intent_type = "explain"
		end
	end

	-- Normalize confidence to 0-1 range (cap at reasonable max)
	local confidence = math.min(winner.score / 2, 1.0)

	-- Check if exploration is needed (full project indexing)
	local needs_exploration = explore_score >= 0.9

	---@type Intent
	local intent = {
		type = intent_type,
		confidence = confidence,
		needs_project_context = proj_score > 0.5 or needs_exploration,
		needs_brain_context = intent_type == "ask" or intent_type == "explain",
		needs_exploration = needs_exploration,
		keywords = winner.keywords,
	}

	return intent
end

--- Get prompt type for system prompt selection
---@param intent Intent Detected intent
---@return string Prompt type for prompts.system
function M.get_prompt_type(intent)
	local mapping = {
		ask = "ask",
		explain = "ask", -- Uses same prompt as ask
		generate = "code_generation",
		refactor = "refactor",
		document = "document",
		test = "test",
	}
	return mapping[intent.type] or "ask"
end

--- Check if intent requires code output
---@param intent Intent
---@return boolean
function M.produces_code(intent)
	local code_intents = {
		generate = true,
		refactor = true,
		document = true, -- Documentation is code (comments)
		test = true,
	}
	return code_intents[intent.type] or false
end

return M

File: agent/intent.py
"""
Intent classification module.

This module is responsible for understanding what the user wants to do.
It replaces all intent classification logic previously in Lua.

Responsibilities:
- Classify user prompts into intent types (ask, code, refactor, etc.)
- Score confidence in the classification
- Detect ambiguity and generate clarification questions
- Never guess when uncertain - ask for clarification

Migrated from:
- lua/codetyper/core/intent/init.lua
- lua/codetyper/features/ask/intent.lua
- lua/codetyper/params/agents/intent.lua
"""

import re
from typing import List, Dict, Tuple, Optional
from .schemas import IntentRequest, IntentResponse, IntentType


# Keyword patterns for each intent type
INTENT_PATTERNS: Dict[IntentType, List[str]] = {
    IntentType.ASK: [
        r"\b(what|why|how|when|where|who|which|explain|describe|tell me|help me understand)\b",
        r"\bis\s+(it|this|that)\b",
        r"\?$",
        r"\b(mean|meaning|purpose|difference|between)\b",
    ],
    IntentType.CODE: [
        r"\b(write|create|generate|implement|add|make|build)\b",
        r"\b(function|class|method|module|component|api|endpoint)\b",
        r"\b(new|from scratch)\b",
    ],
    IntentType.REFACTOR: [
        r"\b(refactor|restructure|reorganize|clean up|improve|simplify)\b",
        r"\b(extract|split|merge|combine|move)\b",
        r"\b(better|cleaner|more readable|more maintainable)\b",
    ],
    IntentType.FIX: [
        r"\b(fix|bug|error|issue|problem|broken|wrong|incorrect)\b",
        r"\b(doesn't work|not working|fails|failing|crash)\b",
        r"\b(debug|troubleshoot|solve)\b",
    ],
    IntentType.DOCUMENT: [
        r"\b(document|documentation|docstring|comment|jsdoc|typedoc)\b",
        r"\b(add comments|write docs|explain the code)\b",
        r"\b(readme|changelog)\b",
    ],
    IntentType.EXPLAIN: [
        r"\b(explain|walk me through|how does|what does)\b",
        r"\b(this code|this function|this class)\b",
        r"\b(step by step|in detail)\b",
    ],
    IntentType.TEST: [
        r"\b(test|tests|testing|spec|specs)\b",
        r"\b(unit test|integration test|e2e|end.to.end)\b",
        r"\b(coverage|assert|expect|mock)\b",
    ],
}

# Keywords that strongly indicate specific intents
STRONG_INDICATORS: Dict[str, IntentType] = {
    "write a function": IntentType.CODE,
    "create a class": IntentType.CODE,
    "implement": IntentType.CODE,
    "refactor": IntentType.REFACTOR,
    "fix the bug": IntentType.FIX,
    "fix this": IntentType.FIX,
    "add tests": IntentType.TEST,
    "write tests": IntentType.TEST,
    "add documentation": IntentType.DOCUMENT,
    "add comments": IntentType.DOCUMENT,
    "explain this": IntentType.EXPLAIN,
    "what does this": IntentType.EXPLAIN,
    "how does this": IntentType.EXPLAIN,
}

# Ambiguous phrases that need clarification
AMBIGUOUS_PATTERNS = [
    r"\bthis\b(?!\s+(code|function|class|file|method))",  # "this" without specifier
    r"\bit\b",  # generic "it"
    r"\b(improve|better|update)\b",  # vague action verbs
    r"\b(change|modify)\b(?!\s+to)",  # change without target
]


class IntentClassifier:
    """
    Classifies user prompts into actionable intents.

    The classifier prioritizes precision over recall - it's better
    to ask for clarification than to misunderstand the user.
    """

    # Minimum confidence to return a classification without clarification
    CONFIDENCE_THRESHOLD = 0.7

    # Weights for different signal sources
    WEIGHTS = {
        "strong_indicator": 0.5,
        "pattern_match": 0.3,
        "context_signal": 0.2,
    }

    def __init__(self):
        """Initialize the classifier."""
        pass

    def classify(self, request: IntentRequest) -> IntentResponse:
        """
        Classify a user prompt into an intent.

        Args:
            request: The classification request with context and prompt

        Returns:
            IntentResponse with classified intent and confidence
        """
        prompt = request.prompt.lower().strip()
        context = request.context.lower() if request.context else ""

        # Check for strong indicators first
        intent, confidence = self._check_strong_indicators(prompt)
        if intent and confidence >= self.CONFIDENCE_THRESHOLD:
            return IntentResponse(
                intent=intent,
                confidence=confidence,
                reasoning=f"Strong indicator phrase detected",
                needs_clarification=False,
            )

        # Score all intent types
        scores = self._score_intents(prompt, context)

        # Find best match
        best_intent = max(scores, key=scores.get)
        best_score = scores[best_intent]

        # Check for ambiguity
        is_ambiguous, ambiguity_reasons = self._check_ambiguity(prompt, scores)

        # Generate reasoning
        reasoning = self._generate_reasoning(best_intent, scores, prompt)

        # Check if we need clarification
        needs_clarification = best_score < self.CONFIDENCE_THRESHOLD or is_ambiguous
        questions = []

        if needs_clarification:
            questions = self._generate_clarification(prompt, scores, ambiguity_reasons)

        return IntentResponse(
            intent=best_intent,
            confidence=best_score,
            reasoning=reasoning,
            needs_clarification=needs_clarification,
            clarification_questions=questions,
        )

    def _check_strong_indicators(self, prompt: str) -> Tuple[Optional[IntentType], float]:
        """Check for strong indicator phrases."""
        for phrase, intent in STRONG_INDICATORS.items():
            if phrase in prompt:
                return intent, 0.9
        return None, 0.0

    def _score_intents(self, prompt: str, context: str) -> Dict[IntentType, float]:
        """Score each intent type based on patterns and context."""
        scores = {intent: 0.0 for intent in IntentType if intent != IntentType.UNKNOWN}

        for intent, patterns in INTENT_PATTERNS.items():
            pattern_score = 0.0
            matches = 0

            for pattern in patterns:
                if re.search(pattern, prompt, re.IGNORECASE):
                    matches += 1
                    pattern_score += 1.0 / len(patterns)

            # Boost if multiple patterns match
            if matches > 1:
                pattern_score = min(pattern_score * 1.2, 1.0)

            scores[intent] = pattern_score

        # Normalize scores
        total = sum(scores.values())
        if total > 0:
            for intent in scores:
                scores[intent] = scores[intent] / total

        # Apply minimum floor to prevent 0 scores
        for intent in scores:
            scores[intent] = max(scores[intent], 0.05)

        # Re-normalize after floor
        total = sum(scores.values())
        for intent in scores:
            scores[intent] = scores[intent] / total

        return scores

    def _check_ambiguity(
        self,
        prompt: str,
        scores: Dict[IntentType, float],
    ) -> Tuple[bool, List[str]]:
        """Check if the prompt is ambiguous."""
        reasons = []

        # Check for ambiguous patterns
        for pattern in AMBIGUOUS_PATTERNS:
            if re.search(pattern, prompt, re.IGNORECASE):
                reasons.append(f"Vague reference detected")
                break

        # Check if top scores are too close
        sorted_scores = sorted(scores.values(), reverse=True)
        if len(sorted_scores) >= 2:
            top_diff = sorted_scores[0] - sorted_scores[1]
            if top_diff < 0.15:
                reasons.append("Multiple intents seem equally likely")

        # Check if prompt is too short
        word_count = len(prompt.split())
        if word_count < 3:
            reasons.append("Prompt is very short")

        return len(reasons) > 0, reasons

    def _generate_reasoning(
        self,
        intent: IntentType,
        scores: Dict[IntentType, float],
        prompt: str,
    ) -> str:
        """Generate human-readable reasoning for the classification."""
        score = scores[intent]

        if score > 0.8:
            return f"Strong match for {intent.value} intent based on keywords and patterns"
        elif score > 0.6:
            return f"Likely {intent.value} intent, though some ambiguity exists"
        elif score > 0.4:
            return f"Possibly {intent.value} intent, but confidence is low"
        else:
            return f"Weak signal for {intent.value} intent, clarification recommended"

    def _generate_clarification(
        self,
        prompt: str,
        scores: Dict[IntentType, float],
        ambiguity_reasons: List[str],
    ) -> List[str]:
        """Generate clarification questions when intent is ambiguous."""
        questions = []

        # Get top 2 intents
        sorted_intents = sorted(scores.items(), key=lambda x: x[1], reverse=True)
        top_intents = [i[0] for i in sorted_intents[:2]]

        # Ask about the distinction
        if IntentType.CODE in top_intents and IntentType.REFACTOR in top_intents:
            questions.append(
                "Do you want to write new code or refactor existing code?"
            )
        elif IntentType.ASK in top_intents and IntentType.EXPLAIN in top_intents:
            questions.append(
                "Are you asking a general question or do you want an explanation of specific code?"
            )
        elif IntentType.FIX in top_intents and IntentType.REFACTOR in top_intents:
            questions.append(
                "Is there a bug to fix, or do you want to improve the code structure?"
            )

        # Generic questions based on ambiguity
        if "Vague reference detected" in ambiguity_reasons:
            questions.append(
                "Could you specify which file, function, or code section you're referring to?"
            )

        if "Prompt is very short" in ambiguity_reasons:
            questions.append(
                "Could you provide more details about what you'd like me to do?"
            )

        # Default question if none generated
        if not questions:
            questions.append(
                "I'm not sure I understand. Could you clarify what you'd like me to do?"
            )

        return questions[:3]  # Limit to 3 questions


# Convenience function for direct classification
def classify_intent(context: str, prompt: str, files: List[str] = None) -> IntentResponse:
    """
    Classify a prompt into an intent.

    Args:
        context: Buffer/file context
        prompt: User's prompt
        files: Referenced file paths

    Returns:
        IntentResponse with classification
    """
    classifier = IntentClassifier()
    request = IntentRequest(
        context=context,
        prompt=prompt,
        files=files or [],
    )
    return classifier.classify(request)
